Src/WindowsServer/WindowsServer.Shared/Implementation/AzureComputeMetadataHeartbeatPropertyProvider.cs

```diff
@@ -0,0 +1,114 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+    internal class AzureComputeMetadataHeartbeatPropertyProvider
+    {
+        internal const string HeartbeatPropertyPrefix = "azInst_"; // to ensure no collisions with base heartbeat properties
+
+        /// <summary>
+        /// Expected fields extracted from Azure IMS to add to the heartbeat properties. 
+        /// Set as internal for testing.
+        /// </summary>
+        internal readonly IReadOnlyCollection<string> ExpectedAzureImsFields = new string[]
+        {
+            "location",
+            "name",
+            "offer",
+            "osType",
+            "placementGroupId",
+            "platformFaultDomain",
+            "platformUpdateDomain",
+            "publisher",
+            "resourceGroupName",
+            "sku",
+            "subscriptionId",
+            "tags",
+            "version",
+            "vmId",
+            "vmSize"
+        };
+
+        /// <summary>
+        /// Flags that will tell us whether or not Azure VM metadata has been attempted to be gathered or not, and
+        /// if we should even attempt to look for it in the first place. 
+        /// </summary>
+        private bool isAzureMetadataCheckCompleted = false;
+
+        private IAzureMetadataRequestor azureInstanceMetadataRequestor = null;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureComputeMetadataHeartbeatPropertyProvider"/> class.
+        /// </summary>
+        public AzureComputeMetadataHeartbeatPropertyProvider() : this(null, false)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureComputeMetadataHeartbeatPropertyProvider"/> class.
+        /// </summary>
+        /// <param name="azureInstanceMetadataHandler">For testing: Azure metadata request handler to use when requesting data from azure specifically. If left as null, an instance of AzureMetadataRequestor is used.</param>
+        /// <param name="resetCheckCompleteFlag">For testing: set to true to reset the check that we've already acquired this data.</param>
+        internal AzureComputeMetadataHeartbeatPropertyProvider(IAzureMetadataRequestor azureInstanceMetadataHandler = null, bool resetCheckCompleteFlag = false)
+        {
+            this.azureInstanceMetadataRequestor = azureInstanceMetadataHandler;
+            if (this.azureInstanceMetadataRequestor == null)
+            {
+                this.azureInstanceMetadataRequestor = new AzureMetadataRequestor();
+            }
+
+            if (resetCheckCompleteFlag)
+            {
+                this.isAzureMetadataCheckCompleted = false;
```

Comment:
I don't see the point of this. The `isAzureMetadataCheckCompleted` field is already false. Is it supposed to be a static field?

Reply:
Ah yes, leftover cruft code that I'd missed. That was indeed a static field at one point that I needed to reset. However, due to the usage pattern of this module I don't believe the use of the static field is altogether prudent.